src/call.rs

```diff
@@ -0,0 +1,120 @@
+//! TODO
+//!
+//!
+
+use crate::ffi_ptr_ext::FfiPtrExt as _;
+use crate::types::{PyAnyMethods as _, PyDict, PyString, PyTuple};
+use crate::{ffi, Borrowed, Bound, IntoPyObjectExt as _, Py, PyAny, PyResult};
+
+pub(crate) mod private {
+    use super::*;
+
+    pub trait Sealed {}
+
+    impl Sealed for () {}
+    impl Sealed for Bound<'_, PyTuple> {}
+    impl Sealed for Py<PyTuple> {}
+
+    pub struct Token;
+}
+
+/// TODO
+pub trait PyCallArgs<'py>: Sized + private::Sealed {

```

Comment:
We should probably add `#[diagnostic::on_unimplemented]` here to hint to users what to do to upgrade if they happen to be using a custom type as args. (Or in general just to hint what to pass in.)

Reply:
That's a good idea, will do that 👍